### ISSUE TITLE
AbbreviationAsWordInName: Numbers break Abbreviation-Scan

### ISSUE THREAD

**Issue by scrayos** [checkstyle/checkstyle#6694]
```
/var/tmp $ javac PlatformEntryPoint.java
**no output**

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">

        <module name="AbbreviationAsWordInName">
            <property name="allowedAbbreviationLength" value="1"/>
            <property name="allowedAbbreviations" value="JC,SC,SLF4J,GUI,IO,TPS"/>
            <property name="ignoreFinal" value="false"/>
            <property name="ignoreStatic" value="true"/>
            <property name="ignoreOverriddenMethods" value="true"/>
        </module>

    </module>
</module>

/var/tmp $ cat PlatformEntryPoint.java
package net.justchunks.jcinclude.object.internal;

/**
 * Working javadoc.
 */
public interface PlatformEntryPoint {

    /**
     * Another working java-doc.
     *
     * @return Local SLF4J-Handler.
     */
    String getSLF4JLogger();
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-8.20-all.jar \
                    -c config.xml PlatformEntryPoint.java
Starting audit...
[WARN] ../PlatformEntryPoint.java:13: Abbreviation in name 'getSLF4JLogger' 
must contain no more than '2' consecutive capital letters. [AbbreviationAsWordInName]
Audit done.
```

---------------

I'd expect to be able to allow abbreviations containing numbers through the allowedAbbreviations" property of the "AbbreviationAsWordInName"-Check.

--------------  

there happens to be a lot of abbreviations in my use case, that contain numbers. The most famous would be "SLF4J", which I tried to whitelist through the [property](http://checkstyle.sourceforge.net/config_naming.html#AbbreviationAsWordInName_Properties) "allowedAbbreviations". This did not work and once I looked into the [responsible check](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AbbreviationAsWordInNameCheck.java#L376) it became clear why that is.

In my Opinion, Numbers should neither start nor end the "abbreviation counter", but they should continue it. That would mean:
"myMETH0DName" would be counted as 7, 
"my4THMethod" would be counted as 3,
"goodN8" would be counted as 2.

I'll prepare a PR to align this with what I would expect, but of course I am open to other opinions regarding this method. The current behavior however, is totally unexpected for me, even though from a "technical" perspective it may makes sense.

**Comment by rnveach** [comment]
According to the google style guide, there is nothing that states numbers are to be included or not included.
http://checkstyle.sourceforge.net/styleguides/google-java-style-20170228.html#s5.3-camel-case

Some googling,
https://stackoverflow.com/questions/37496068/how-to-deal-with-numbers-in-camelcase-naming-policy
> playSize32, playUrl24M4a
> you should try to avoid variables names containing them

@Scrayos 
> Numbers should neither start nor end the "abbreviation counter", 
> "goodN8" would be counted as 2

Doesn't this violate your rules that numbers shouldn't end the counter. It should be counted as 1 not 2 since N is the only capital and there are no consecutive.

**Comment by rnveach** [comment]
@romani @pbludov What is your take?

**Comment by Scrayos** [comment]
> Doesn't this violate your rules that numbers shouldn't end the counter. It should be counted as 1 not 2 since N is the only capital and there are no consecutive.

My phrasing was probably misleading. What I meant was that numbers/digits should neither set the "beginAbbreviation" variable to true nor false. That means while it will never be the first char of an abbreviation, it could be the last. Or to say it in another fashion: Numbers should be Abbreviation-Neutral. If we're currently within an abbreviation, preserve that. If we're not within an abbreviation, preserve that instead.

> It could be argued that numbers should be a word in itself, which makes sense. Its not SLF4J but SLF 4 Java or SLF for Java

I would agree with that, although I'd question if "getSLFForJavaLogger" would really be "better" considering developers expectations regarding the method name.

**Comment by romani** [comment]
@Scrayos , please follow required issue template, it is MUST for all (as it allow to avoid bunch of questions to issue reporter). Please follow.
I need to see your exact config file and exact variables names that you think should not raise violation.

**Comment by Scrayos** [comment]
@romani I added the stuff from the issue template, although I'm pretty sure I already outlined really thoroughly what the problem is and I even showed it within the code. Anyways, I understand your need for standardized issues, so I've added it.

**Comment by romani** [comment]
> checkstyle-X.XX-all.jar

@Scrayos , please suggest what we should update in termplate to make it clear that here we need real jar, that show a version. It is critical point in issue template. Half pf issue self resolved ones user become to understand that they use old version of checkstyle library in their plugins/integration.
Please update description with version.

**Comment by Scrayos** [comment]
@romani I added the name. I directly downloaded the jar before submitting the issue and then again before I added the issue template. So I am definitely on the newest version.

Regarding the issue template, I think a comment at the start of the template, that states "also the commands should be replaced with the real executed command" or the likes would be sufficient. And I'd manually request the version, as some people may rename the jar.

**Comment by romani** [comment]
Issue description is updated. It would be much quicker to reply to you if you follow template.

I would be awesome if you explain why you put for `cat config.xml` link to config. When excepted full content of config.

Case that you provide is valid issue.

The way on how to resolve it, should be discussed. Treating of numbers in some way might be very questionable and some users might look on names differently. 

supportsIpv6OnIos - supports Ipv6 On Ios
supportsIPV6OOnIos - supports IPV6O On Ios - should it be violation of not ? if IPV6O is known term in company.

Does it work for you if we teach Check to match allowedAbbreviations to name to allow parsing for subwords ?
so if allowedAbbreviations="IPV6O", then supportsIPV6OOnIos parsed as "supports IPV6O On Ios" is OK, and supportsIPV6KOnIos parsed as "supports IPV 6K On Ios"(or in some other way) is violation

in your case:
allowedAbbreviations="SLF4J", then getSLF4JLogger is parsed as "get SLF4J Logger".

So user do know his code and define kind of vocabular for Check to detect words properly.

@Scrayos , does it make sense ? is there any problems that you might see ?

**Comment by romani** [comment]
We need to be able to to let user have abbreviation as "PIXEL3A" if he likes it( or as temporal compromise on legacy code) in name like `getPIXEL3APrice`.

**Comment by Scrayos** [comment]
> It would be much quicker to reply to you if you follow template.

Yeah, I'm sorry about that. As I already wrote, I was sure I outlined all necessary aspect, as I specifically sent the link to the bug in the code, so I thought for example version information should not be needed, as it could've been understood by the code alone, independent of the version of checkstyle I was using.

> I would be awesome if you explain why you put for cat config.xml link to config. When excepted full content of config.

I did that out of courtesy for everyone's scrolling wheels. The configuration is approx. 1000 Lines long, some lines being so long, that they would've wrapped in the code-area, resulting in even more lines. I decided that the advantage of having everything in one place wouldn't suffice the strain on everyone's eyes, hands and mouses, that embedding the code directly would've had.

> supportsIpv6OnIos - supports Ipv6 On Ios
supportsIPV6OOnIos - supports IPV6O On Ios - should it be violation of not ? if IPV6O is known term in company.

In my opinion the second case should raise a violation, as long as `IPV6O` isn't a defined term for `allowedAbbreviations` in the configuration.

> Does it work for you if we teach Check to match allowedAbbreviations to name to allow parsing for subwords ?
so if allowedAbbreviations="IPV6O", then supportsIPV6OOnIos parsed as "supports IPV6O On Ios" is OK, and supportsIPV6KOnIos parsed as "supports IPV 6K On Ios"(or in some other way) is violation

What about words that end with the number? Like `getSLF4Handler` (without the J) – would they get recognized as subwords: "get slf4 handler" or would they get treated this way: "get slf 4handler"?

> in your case:
allowedAbbreviations="SLF4J", then getSLF4JLogger is parsed as "get SLF4J Logger".

Yeah, that would be suitable to my use case. I wonder about the case I outlined above, however. The behavior you wrote here would be what I expect for my error regarding `SLF4J`.

> So user do know his code and define kind of vocabular for Check to detect words properly.

I am worried that the adjustment of the check you proposed could look kinda arbitrary to users that are not aware of this ticket. That is why I suggested that numbers should just silently inherit/keep the casing of their preceding letter. So here: `getSLF4JLogger` the `4` is treated as if it was an uppercase letter, while here: `getSlf4jLogger` it would be interpreted as if it was a lowercase letter. This would also require only a very tiny adjustment, that I submitted within [this pull-request](https://github.com/checkstyle/checkstyle/pull/6695). This – at least in my opinion – is a consistent behavior that users of checkstyle could rely on. But I may be biased of course.

> We need to be able to to let user have abbreviation as "PIXEL3A" if he likes it( or as temporal compromise on legacy code) in name like getPIXEL3APrice.

That is exactly what this issue is about.

**Comment by romani** [comment]
> What about words that end with the number? Like getSLF4Handler (without the J) – would they get recognized as subwords: "get slf4 handler" or would they get treated this way: "get slf 4handler"?

user need to define words in `allowedAbbreviations` we should use them as strict match.
if user define `allowedAbbreviations=SLF4J` and name is `getSLF4Handler` it will be violation, even `getSLF4JHandler` be violation as we will parse it as "get SLF4 JHandler".

> I am worried that the adjustment of the check you proposed could look kinda arbitrary to users that are not aware of this ticket

we will not change code only, we will update documentation and provide examples.

> that I submitted within this pull-request. This – at least in my opinion – is a consistent behavior that users of checkstyle could rely on. But I may be biased of course.

I understand you, but in my experience, users come with such interesting cases, that we even can not predict. Naming in legacy code could be weird and unchangeable. Some uncertain gut feeling tell me to not make a decision `numbers should just silently inherit/keep the casing of their preceding letter.` for user, without options to disable it.

"log2Db", "log2db", "LOG2Db", "LoG2DB" 
"SLF4Addapter", "SLF4JAddapter", "SLF4J2MyLoggerAddpater"

It is reasonable that number could be a used as:
- suffix in name (example: "DATA360")
- part of word (example: "getPIXEL3APrice", "getSLF4JLogger")
- as word (example: "DATA2DateConverter", "DATA4Me").
so your ideas is very good.

but "getSLF4JLogger" is "get SLF4 JLogger" or "get SLF4J Logger" ?
if user have `allowedAbbreviations=JDK` and name is "getSLF4JDKLogger", by your logic it will be "get SLF4JDK Logger", so violation and user will change it  to "getSlf4JDKLogger", all good.
if user have `allowedAbbreviations=SLF,JDK` and name is "getSLF4JDKLogger", by your logic it will be "get SLF4JDK Logger", so violation, BUT it should might treated/read by user as "get SLF 4 JDK Logger" !!!!!

We need to give user ability to control this behavior a bit, or on most weird cases, we able to suppress them:
 - Suppression could be done by https://checkstyle.org/config_filters.html#SuppressionFilter (see option to do it by regexp on "message"). 
- control of parsing is only by some options in Check. Your approach is easy but have some issues. My approach is more complicated to implement, but give bit more to user. ..... For sub words as SLF and SLF4J we can warn user that we follow order in allowedAbbreviations, first match finish word recognition, so if user need both  SLF and SLF4J. He need to do define `allowedAbbreviations=SLF4J,SLF`.

@Scrayos , any other ideas ? corrections ? nuances ?

**Comment by Scrayos** [comment]
Sounds good, but I'm afraid I won't be able to contribute any code for this, as this more complex implementation would require more time than I can spare at the moment. Your solution sounds nice from a user-side (which is the most important perspective), but would require extensive testing and care to minimize any performance-drawbacks (introduced by for example regex' backtracking). I am overall of the opinion that this is worth the effort though.

Thanks for your extensive feedback on this and for shedding light on all the different aspects that would not be covered by my PR.

**Comment by romani** [comment]
> care to minimize any performance-drawbacks 

user care about reliable results first of all not a performance, you will not damage performance that much. 

> but would require extensive testing 

it is un-avoidable in all cases.

> but I'm afraid I won't be able to contribute any code for this, as this more complex implementation would require more time than I can spare at the moment. 

yes, better solutions always takes time.
meanwhile I would suggest to use suppression for such cases.

I hope you will find time to make such update, it should not very complicated if we forget a goal to be super performant. If in future ..... when such Check become so popular and it at top of our performance problems  .... somebody will improve it ( functionality will stay the same ).

**Comment by stechio** [comment]
Since today I stumbled upon this problem, I thought to report my test case to revive this 3-year-old issue.

NOTE: despite abbreviations here below are obfuscated into abstract alphanumeric sequences, my test case represents a real-world use case, NOT a theoretical edge case.

---

Check documentation: https://checkstyle.org/config_naming.html#AbbreviationAsWordInName

```bash
$ javac Test.java

$ cat test-checkstyle.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="AbbreviationAsWordInName">
      <property name="allowedAbbreviations" value="ABCDEFG,ABCDE12,ABCDE"/>
    </module>
  </module>
</module>

$ cat Test.java
public class Test {
  static class ABCDEFGRunner {
    void run() {
    }
  }

  static class ABCDE12Runner {
    void run() {
    }
  }

  static class ABC12345Runner {
    void run() {
    }
  }

  public static void main(String[] args) {
    new ABCDE12Runner().run();
  }
}

$ RUN_LOCALE="-Duser.language=en -Duser.country=US"
$ java $RUN_LOCALE -jar checkstyle-10.1-all.jar -c test-checkstyle.xml Test.java
Starting audit...
[ERROR] /home/stechio/Projects/checkstyle-test/temp/Test.java:7:16: Abbreviation in name 'ABCDE12Runner' must contain no more than '4' consecutive capital letters. [AbbreviationAsWordInName]
Audit done.
Checkstyle ends with 1 errors.
```

---

To my understanding:
- "ABC12345Runner" PASSES because its relevant abbreviation is only "ABC" and `allowedAbbreviationLength` is 3 by default.
- "ABCDEFGRunner" PASSES because, despite its relevant abbreviation "ABCDEFG" exceeds `allowedAbbreviationLength`, it is configured in `allowedAbbreviations`.
- "ABCDE12Runner" FAILS because, despite being configured in `allowedAbbreviations` (both full "ABCDE12" and letter-only "ABCDE"), the parser cannot properly handle such a sequence.

EXPECTED BEHAVIOR
Aside from your complex reasoning in this thread (which would require to extend the current implementation), I think my case shows there's definitively a bug in the current implementation: "ABCDE12Runner" should pass because `allowedAbbreviations` contains both "ABCDE12" and "ABCDE" -- one of the two may not work (because of the trailing digits), but at least the other one should pass, shouldn't it?
